aea/configurations/components.py

```diff
@@ -0,0 +1,178 @@
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""This module contains definitions of agent components."""
+import inspect
+import logging
+import re
+import types
+from abc import ABC
+from pathlib import Path
+from typing import Dict, Optional, Type, cast
+
+from aea.configurations.base import (
+    ComponentConfiguration,
+    ComponentId,
+    ComponentType,
+    ConnectionConfig,
+    PublicId,
+)
+from aea.helpers.base import _SysModules, load_init_modules, load_module
+
+logger = logging.getLogger(__name__)
+
+
+def component_class_from_type(component_type: ComponentType) -> Type["Component"]:
+    """Get component class from component type."""
+    from aea.protocols.base import Protocol
+    from aea.connections.base import Connection
+    from aea.skills.base import Skill
+
+    if component_type == ComponentType.PROTOCOL:
+        return Protocol
+    elif component_type == ComponentType.CONNECTION:
+        return Connection
+    elif component_type == ComponentType.SKILL:
+        return Skill
+    elif component_type == ComponentType.CONTRACT:
+        # TODO
+        raise NotImplementedError
+    else:
+        raise ValueError
+
+
+class Component(ABC):
+    """Abstract class for an agent component."""
+
+    def __init__(self, configuration: ComponentConfiguration, is_vendor: bool = False):
+        """
+        Initialize a package.
+
+        :param configuration: the package configuration.
+        :param is_vendor: whether the package is vendorized.
+        """
+        self._configuration = configuration
+        self._directory = None  # type: Optional[Path]
+        self._is_vendor = is_vendor
+
+        # mapping from import path to module object
+        # the keys are dotted paths of Python modules.
+        self.importpath_to_module = {}  # type: Dict[str, types.ModuleType]
+
+    @property
+    def component_type(self) -> ComponentType:
+        """Get the component type."""
+        return self._configuration.component_type
+
+    @property
+    def is_vendor(self) -> bool:
+        """Get whether the component is vendorized or not."""
+        return self._is_vendor
+
+    @property
+    def prefix_import_path(self):
+        """Get the prefix import path for this component."""
+        if self.is_vendor:
+            return "packages.{}.{}.{}".format(
+                self.public_id.author,
```

Comment:
should we forbid some author strings that could make some conflicts with the import paths below (non-vendor case)? E.g. `skills`, `protocols`...

Reply:
Yes definitely! :)

Reply:
Open issue

Reply:
https://github.com/fetchai/agents-aea/issues/961